File: academic-system/src/main/java/org/example/academic/system/repository/TxtRepository.java
package org.example.academic.system.repository;

import org.example.academic.system.model.*;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class TxtRepository implements PersistenceStrategy {

    @Override
    public void save(List<AcademicClass> classes) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter("academic_data.txt"))) {
            for (AcademicClass academicClass : classes) {
                writer.write("Class Code: " + academicClass.getCode() + "\n");
                writer.write("Class Title: " + academicClass.getTitle() + "\n");
                writer.write("Assessments:\n");

                if (academicClass.getAssessments() != null) {
                    for (Assessment assessment : academicClass.getAssessments()) {
                        writer.write("  - Type: " + assessment.getType() +
                                " | Value: " + assessment.getValue() +
                                " | Weight: " + assessment.getWeight() + "\n");
                    }
                }
                writer.write("=========================================\n");
            }
        } catch (Exception e) {
            System.err.println("Erro ao salvar os dados em TXT: " + e.getMessage());
            e.printStackTrace();
        }
    }

    @Override
    public List<AcademicClass> load() {
        List<AcademicClass> classes = new ArrayList<>();
        File file = new File("academic_data.txt");

        // Se o arquivo ainda não existir (primeira execução), retorna a lista vazia
        if (!file.exists()) {
            return classes;
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
            String line;
            AcademicClass currentClass = null;

            while ((line = reader.readLine()) != null) {
                line = line.trim();

                if (line.startsWith("Class Code:")) {
                    String code = line.replace("Class Code:", "").trim();
                    String titleLine = reader.readLine();
                    String title = titleLine != null ? titleLine.replace("Class Title:", "").trim() : "";

                    // Recria a turma usando o construtor do Carlos Pereira
                    currentClass = new AcademicClass(code, title);
                    classes.add(currentClass);

                } else if (line.startsWith("- Type:") && currentClass != null) {
                    // Quebra a linha da avaliação para extrair os dados
                    // Exemplo: - Type: exam | Value: 6.0 | Weight: 10.0
                    String type = line.substring(line.indexOf("Type:") + 5, line.indexOf("|")).trim();

                    String valuePart = line.substring(line.indexOf("Value:") + 6, line.indexOf("|", line.indexOf("Value:"))).trim();
                    double value = Double.parseDouble(valuePart);

                    String weightPart = line.substring(line.indexOf("Weight:") + 7).trim();
                    double weight = Double.parseDouble(weightPart);

                    // Instancia a classe filha correta com base no tipo guardado
                    if (type.equalsIgnoreCase("exam")) {
                        currentClass.addAssessment(new Exam(value, weight));
                    } else if (type.equalsIgnoreCase("practical") || type.equalsIgnoreCase("Practical Assignment")) {
                        currentClass.addAssessment(new PracticalAssignment(value, weight));
                    } else if (type.equalsIgnoreCase("assignment")) {
                        currentClass.addAssessment(new Assignment(value, weight));
                    } else if (type.equalsIgnoreCase("seminar")) {
                        currentClass.addAssessment(new Seminar(value, weight));
                    }
                }
            }
            System.out.println("Dados carregados do TXT com sucesso! Total de turmas: " + classes.size());
        } catch (Exception e) {
            System.err.println("Erro ao carregar os dados do TXT: " + e.getMessage());
        }
        return classes;
    }

    @Override
    public String getFormatName() {
        return "TXT";
    }
}

File: academic-system-python/academic_system/repository/txt_repository.py
import os
from typing import List

from .persistence_strategy import PersistenceStrategy
from ..model.academic_class import AcademicClass
from ..model.assignment import Assignment
from ..model.exam import Exam
from ..model.practical_assignment import PracticalAssignment
from ..model.seminar import Seminar


class TxtRepository(PersistenceStrategy):
    FILE_NAME = "academic_data.txt"

    # Escreve cada turma e suas avaliações no arquivo TXT em formato legível
    def save(self, classes: List[AcademicClass]) -> None:
        try:
            with open(self.FILE_NAME, "w", encoding="utf-8") as f:
                for academic_class in classes:
                    f.write(f"Class Code: {academic_class.code}\n")
                    f.write(f"Class Title: {academic_class.title}\n")
                    f.write("Assessments:\n")
                    for assessment in academic_class.assessments:
                        f.write(
                            f"  - Type: {assessment.type}"
                            f" | Value: {assessment.value}"
                            f" | Weight: {assessment.weight}\n"
                        )
                    f.write("=========================================\n")
        except Exception as e:
            print(f"Erro ao salvar os dados em TXT: {e}")

    # Lê o arquivo TXT linha a linha e reconstrói as turmas com suas avaliações
    def load(self) -> List[AcademicClass]:
        classes = []
        if not os.path.exists(self.FILE_NAME):
            return classes
        try:
            with open(self.FILE_NAME, "r", encoding="utf-8") as f:
                current_class = None
                for line in f:
                    line = line.strip()
                    if line.startswith("Class Code:"):
                        code = line.replace("Class Code:", "").strip()
                        title_line = next(f, "").strip()
                        title = title_line.replace("Class Title:", "").strip()
                        current_class = AcademicClass(code, title)
                        classes.append(current_class)
                    elif line.startswith("- Type:") and current_class:
                        type_part = line[line.index("Type:") + 5:line.index("|")].strip()
                        value_part = line[line.index("Value:") + 6:line.index("|", line.index("Value:"))].strip()
                        weight_part = line[line.index("Weight:") + 7:].strip()
                        value = float(value_part)
                        weight = float(weight_part)
                        current_class.add_assessment(self._create_assessment(type_part, value, weight))
        except Exception as e:
            print(f"Erro ao carregar os dados do TXT: {e}")
        return classes

    # Instancia a subclasse correta de Assessment com base no tipo lido do TXT
    def _create_assessment(self, type_str: str, value: float, weight: float):
        t = type_str.lower()
        if t == "exam":
            return Exam(value, weight)
        elif t in ("practical", "practical assignment"):
            return PracticalAssignment(value, weight)
        elif t == "seminar":
            return Seminar(value, weight)
        return Assignment(value, weight)

    # Retorna o nome do formato para uso nos logs e relatórios
    def get_format_name(self) -> str:
        return "TXT"
